File: MQTT_main_server/Redis_Lcard/MQTTMessage.cpp
#include "MQTTMessage.h"
#include <sstream>

MQTTMessage::MQTTMessage(const std::string &topic,
                         const std::string &command,
                         const std::map<std::string, std::string> &params,
                         const std::string &device,
                         double serverTimestamp,
                         const std::string &clientResponse,
                         double serverReceivedTimestamp)
    : topic(topic),
    command(command),
    params(params),
    device(device),
    serverTimestamp(serverTimestamp),
    clientResponse(clientResponse),
    serverReceivedTimestamp(serverReceivedTimestamp) {}

std::string MQTTMessage::toJSON() const {
    json message = {
        {"command", command},
        {"params", params},
        {"device", device},
        {"server_timestamp", serverTimestamp},
        {"client_response", clientResponse},
        {"server_received_timestamp", serverReceivedTimestamp}
    };
    return message.dump();
}

void MQTTMessage::publish(MQTTHandler &mqttClient) const {
    mqttClient.publish(topic, toJSON());
}

MQTTMessage MQTTMessage::fromJSON(const std::string &topic, const std::string &jsonData) {
    try {
        auto data = json::parse(jsonData);
        return MQTTMessage(
            topic,
            data.at("command").get<std::string>(),
            data.value("params", std::map<std::string, std::string>{}),
            data.value("device", ""),
            data.value("server_timestamp", std::time(nullptr)),
            data.value("client_response", ""),
            data.value("server_received_timestamp", 0.0)
            );
    } catch (const std::exception &e) {
        throw std::runtime_error(std::string("JSON parsing error: ") + e.what());
    }
}

void MQTTMessage::setClientResponse(const std::string &response, double serverReceivedTimestamp) {
    clientResponse = response;
    this->serverReceivedTimestamp = serverReceivedTimestamp;
}

std::string MQTTMessage::toString() const {
    std::ostringstream oss;
    oss << "MQTTMessage(topic=" << topic
        << ", command=" << command
        << ", params={";

    for (const auto &[key, value] : params) {
        oss << key << ":" << value << ", ";
    }

    oss << "}, device=" << device
        << ", server_timestamp=" << serverTimestamp
        << ", client_response=" << clientResponse
        << ", server_received_timestamp=" << serverReceivedTimestamp
        << ")";

    return oss.str();
}

File: MQTT_main_server/Utils/MQTTMessage.py
import json
import time


class MQTTMessage:
    def __init__(self, topic, device=None, time = time.time()):
        """
        Инициализация сообщения MQTT.

        :param topic: MQTT топик для публикации
        :param device: Имя устройства (если необходимо)
        :param time: Время отправки
        """
        self.topic = topic
        self.device = device
        self.time = time

    def to_json(self):
        """Возвращает сообщение в формате JSON."""
        message = {
            "topic": self.topic,
            "device": self.device,
            "time": self.time
        }
        message = json.dumps(message)
        return message

    @staticmethod
    def from_json(json_data):
        """Создает объект MQTTMessage из JSON строки."""
        try:
            data = json.loads(json_data)
            return MQTTMessage(
                topic=data.get("topic"),
                device=data.get("device"),
                time=data.get("time")
            )
        except json.JSONDecodeError as e:
            raise ValueError(f"Ошибка декодирования JSON: {e}")

    def __str__(self):
        """Возвращает строковое представление сообщения."""
        return (f"MQTTMessage(topic={self.topic}, device={self.device}")
